### PR TITLE
katago: 1.11.0 -> 1.12.4

### PR DESCRIPTION
**Pull request by OmnipotentEntity**
###### Description of changes

Added TensorRT backend, refactored backend selection, IMPORTANT also bumps TensorRT version to 8.5.3 for CUDA 11.7 and 11.8.  (Per Nvidia's documentation, 8.4.0.6 does not officially support 11.7, but we're using for it.  Additionally, KataGo requires 8.5 or better for the new TensorRT backend.  8.4.0.6 remains the default for 11.6 and lower, in interest of not messing things up as much as possible.)

New version also adds support for nested residual block networks, such as the b18 network that was used in the UEC.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by aidalgol — APPROVED** [review]
I can't comment on the Katago derivation, but I like the refactoring you've done in the TensorRT Nix code.

File: pkgs/games/katago/default.nix [hunk 1]
```diff
@@ -85,17 +89,20 @@ stdenv.mkDerivation rec {
...
   installPhase = ''
     runHook preInstall
     mkdir -p $out/bin; cp katago $out/bin;
-  '' + lib.optionalString enableCuda ''
+  '' + lib.optionalString (backend == "cuda") ''
     wrapProgram $out/bin/katago \
       --prefix LD_LIBRARY_PATH : "/run/opengl-driver/lib"
+  '' + lib.optionalString (backend == "tensorrt") ''
+      wrapProgram $out/bin/katago \
+        --prefix LD_LIBRARY_PATH : "/run/opengl-driver/lib"
```
Comment (SuperSandro2000) [inline pkgs/games/katago/default.nix:105]:
```suggestion
  '' + lib.optionalString (backend == "cuda" || backend == "tensorrt") ''
    wrapProgram $out/bin/katago \
      --prefix LD_LIBRARY_PATH : "/run/opengl-driver/lib"
```

File: pkgs/development/libraries/science/math/tensorrt/extension.nix [hunk 2]
```diff
@@ -29,22 +29,89 @@ final: prev: let
...
+  {fullVersion, fileVersionCuda, fileVersionCudnn ? null} :
+  "TensorRT-${fullVersion}.Linux.x86_64-gnu.cuda-${fileVersionCuda}" + lib.optionalString (fileVersionCudnn != null) ".cudnn${fileVersionCudnn}" + ".tar.gz";
+
   tensorRTVersions = {
+    "8.6.0" = [
+      rec {
+        fileVersionCuda = "11.8";
+        fullVersion = "8.6.0.12";
+        sha256 = "sha256-wXMqEJPFerefoLaH8GG+Np5EnJwXeStmDzZj7Nj6e2M=";
+        tarball = tarballURL {inherit fileVersionCuda fullVersion;};
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/science/math/tensorrt/extension.nix:42]:
```suggestion
        tarball = tarballURL { inherit fileVersionCuda fullVersion; };
```

File: pkgs/development/libraries/science/math/tensorrt/extension.nix [hunk 3]
```diff
@@ -29,22 +29,89 @@ final: prev: let
       else throw "tensorrt-${tensorRTDefaultVersion} does not support your cuda version ${cudaVersion}"; };
   in allBuilds // defaultBuild;
 
+  tarballURL =
+  {fullVersion, fileVersionCuda, fileVersionCudnn ? null} :
+  "TensorRT-${fullVersion}.Linux.x86_64-gnu.cuda-${fileVersionCuda}" + lib.optionalString (fileVersionCudnn != null) ".cudnn${fileVersionCudnn}" + ".tar.gz";
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/science/math/tensorrt/extension.nix:34]:
```suggestion
  tarballURL = { fullVersion, fileVersionCuda, fileVersionCudnn ? null }:
    "TensorRT-${fullVersion}.Linux.x86_64-gnu.cuda-${fileVersionCuda}"
    + lib.optionalString (fileVersionCudnn != null) ".cudnn${fileVersionCudnn}"
    + ".tar.gz";
```

File: pkgs/games/katago/default.nix (hunk 1 above)
Comment (OmnipotentEntity) [reply pkgs/games/katago/default.nix]: Good catch.  This was slightly different at some point when I was testing things, and I never noticed that they became the same.

**Review by aidalgol — APPROVED** [review]

**Review by SuperSandro2000 — APPROVED** [review]

**Pull request merged**